pkg/threescale/backend/manager.go

```diff
@@ -212,6 +218,61 @@ func (b *Backend) Flush() {
 			}
 			service[i].authResp = resp
 		}
+
+		for _, cachedApp := range service {
+			app, ok := b.cache.Get(cachedApp.cacheKey)
+			if !ok {
+				continue
+			}
+			app.Lock()
+			if cachedApp.reportingErr {
+				app.Unlock()
+				continue
+			}
+
+			if cachedApp.authErr {
+				app.Unlock()
+				continue
+			}
+			app.LastResponse = cachedApp.authResp
+			// we now want to walk over the cache and check for a diff between the current state
+			// and the state that existed when we took the snapshot originally
+			// this allows us to consider requests that have been made since we released the original lock
```

Comment:
Definitely need to extract some of the code below as the logic is very complex and makes reasoning about this function very difficult.